### DISCUSSION TITLE
[FR] craft\services\Fields::getFieldsByType()

### DISCUSSION THREAD

**Discussion by lindseydiloreto** [craftcms/cms#12381] [Ideas] [answered]
It would be great if there were a simple service method for fetching all fields of a certain **type**...

```php
// Get all Plain Text fields
$plainTextFields = Craft::$app->getFields()->getFieldsByType(PlainText::class);
```

That method doesn't yet exist, so here's what we currently have to do to filter by field type...

```php
// Get all fields
$allFields = Craft::$app->getFields()->getAllFields();

// Keep only Plain Text fields
$plainTextFields = array_filter($allFields, function($field) {
    return is_a($field, PlainText::class);
});
```

I used `PlainText::class` to keep the example above simple.

### Real-World Application

This method would be truly helpful when writing a **migration** for a field type. Writing migrations are already a PITA.

Within the context of the Google Maps plugin, I need to write a migration that targets every single Address field (`AddressField::class`). This method would be super helpful...

```php
$addressFields = Craft::$app->getFields()->getFieldsByType(AddressField::class);
```

**Comment by brandonkelly** [comment, marked as answer]
Why not? Added for 4.4. (3ca0ec5d9170a146005e1ad9d346aa62b90eab51)

**Reply by lindseydiloreto** [reply]
Amazing, thanks @brandonkelly!! 🍺

**Reply by brandonkelly** [reply]
Craft 4.4 is out now!

**Reply by proimage** [reply]
Trying to get this to work in Twig, but I guess I'm passing the wrong string since it's an empty array (and I _do_ have multiple plain text fields):
```
{% set textFields = craft.app.fields.getFieldsByType('plaintext') %}
```
I've tried various variations on capitalization, to no avail. What am I doing wrong, @brandonkelly ?

**Reply by i-just** [reply]
@proimage, you need to call it with the class qualifier, e.g. 
`{% set textFields = craft.app.fields.getFieldsByType('craft\\fields\\PlainText') %}`

**Reply by proimage** [reply]
Ahh, thanks! What would that be for a field from a 3rd-party plugin, such as Neo? Or even better, could you teach me to fish and point out where I'd look to figure the string out for whatever field I'm dealing with? ;)

**Reply by brandonkelly** [reply]
@proimage It’s just always going to be whatever the field type’s full class name is. Same string you’ll find in the `type` column in your `fields` table, or the `type` properties for your field configs in `config/project/`. (The backslashes have to be escaped (`craft\\field\\PlainText` instead of `craft\fields\PlainText`) so Twig interprets them as a backslash, rather than an escape character.)

**Reply by proimage** [reply]
Perfect, thanks y'all!
